src/Database/PostgreSQL/Simple/Transaction.hs

```diff
@@ -157,20 +157,19 @@ withTransactionModeRetry :: TransactionMode -> (SqlError -> Bool) -> Connection
 withTransactionModeRetry mode shouldRetry conn act =
     mask $ \restore ->
         retryLoop $ E.try $ do
-            a <- restore act
+            a <- restore act `E.onException` rollback_ conn
```

Comment:
BTW, `rollback_` is not exported and it's also suspicious. It catches exceptions thrown when issuing the rollback. Why is that a good idea?